include/uart.h

```diff
@@ -64,9 +64,19 @@ extern "C" {
  */
 #define UART_ERROR_BREAK    (1 << 3)
 
+/**
+ * @typedef uart_irq_callback_user_data_t
+ * @brief Define the application callback function signature for
+ * uart_irq_callback_user_data_set() function.
+ *
+ * @param user_data Arbitrary user data.
+ */
+typedef void (*uart_irq_callback_user_data_t)(void *user_data);
```

Comment:
This data type sounds like a user data of a callback, very confusing.

Reply:
> This data type sounds like a user data of a callback, very confusing.

Let's fix it, choose:

1. uart_irq_callback_with_user_data_t
2. uart_irq_callback_w_user_data_t
3. uart_irq_callback_w_ud_t
4. uart_irq_cb_w_ud_t
5. uart_irq_user_data_callback_t
...
20. Your choice.

Reply:
Why change the name of the callback type at all?
uart_irq_callback_t still seems appropriate to me.  It just takes a parameter now.

Reply:
I realize this patch is leaving the older APIs in place, but does it need to?  Let me take another look.

Reply:
@pfalcon To clarify: why are we going to through all this trouble to maintain a legacy version of uart_irq_callback_t (and related functions which use it) when there are no in-tree users and we're changing the UART API for irq_callback_set anyway?  Either way, we're breaking users.

The only way I can see where we don't immediately break users would be to keep the irq_callback_set API call the way it is, and add a new irq_callback_user_data_set API which can be used to set the user_data if needed.  In this case, the exposed function uart_irq_callback_user_data_set() would only take struct device *dev and void *user_data params.

Reply:
> To clarify: why are we going to through all this trouble to maintain a legacy version of uart_irq_callback_t

So, there're to side of things:

1. Public API of UART subsystem (i.e. API between UART subsys and user apps).
2. Private (hopefully) API between UART subsys and UART drivers.

Public API of UART subsystem is in use in the tree, even if uart_irq_callback_t does not occur much explicitly. Users include: uart_pipe, subsys/console, some tests, maybe some samples, your https://github.com/zephyrproject-rtos/zephyr/pull/6344. Then there can be out of tree users too. So, a care is taken to not break that usage, so existing uart_irq_callback_t is left as is (as well as actuall callback funcs with the same signature, even if not declared as uart_irq_callback_t), and a new setter and callback type is introduced for user_data-taking callback.

It's different story with private'ish API between subsys and UART drivers. Maintaining compatibility there would be rather cumbersome, so there we switch to uart_irq_callback_user_data_t right away (and older uart_irq_callback_t is handled as a particular special case of it using "ducktyping").